core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spi;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.reflect.ClassPath;
+import com.google.common.reflect.ClassPath.ClassInfo;
+import org.testng.annotations.Test;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.UncheckedIOException;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import static com.google.common.collect.Sets.difference;
+import static java.lang.ClassLoader.getPlatformClassLoader;
+import static java.lang.ClassLoader.getSystemClassLoader;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestSpiBackwardCompatibility
+{
+    private static final Map<String, Set<String>> BACKWARD_INCOMPATIBLE_CHANGES = ImmutableMap.<String, Set<String>>builder()
+            // When updating this map, please try to remove backward incompatible changes for old versions.
+            // Also consider mentioning backward incompatible changes in release notes.
+            .put("375", ImmutableSet.of(
```

Comment:
This is an example entry which makes adding new entries easier, but this example entry should be removed already (per comment above + 376 is out). Let's make adding new entries easier in some more durable manner, perhaps as a commented-out copy-paste-ready example.

Reply:
I am not sure. 375-SNAPSHOT is almost compatible with 375 which is a good thing.

Also since now version is hardcoded, I no longer think that the key (`"375"`) is needed here.

Reply:
The purpose of key-ing on version was to enforce that backwards incompatible change declaration is not implicitly carried over to the next release.

Let's imagine

- v 100: we have a public method foo
- v 101: we remove the method foo and declare this as a backward incompatible change
- v 102: we reintroduce method foo
- v 103-SNAPSHOT: we remove method foo -- we want the test to fail, as this is a _new_ backward incompatible change; for it to fail, we need to ensure the 101's suppression is not carried forwards implicitly